ldms/src/sampler/zfs_leafvdevs/zfs_leafvdevs.c

```diff
@@ -0,0 +1,449 @@
+        }
+
+        record_instance = ldms_record_alloc(sampler_base->set,
+                                            index_store.vdev_rec_idx);
+
+        if (record_instance == NULL) {
+                log_fn(LDMSD_LDEBUG, SAMP": ldms_record_alloc() failed, resizing metric set\n");
+                resize_metric_set();
+		record_instance = ldms_record_alloc(sampler_base->set,
+				index_store.vdev_rec_idx);
+		if (record_instance == NULL)
+		  rc = 2;
+        }
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+						record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[0],
+				pool_name);
+	        /* zpool state  1 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[1],
+				zpool_state_to_name((vdev_state_t)vs->vs_state, (vdev_aux_t)vs->vs_aux));
+	        /* vdevname     2 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[2],
+						vdev_name);
+	        /* alloc        3 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[3], vs->vs_alloc);
+	        /* free         4 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[4], vs->vs_space - vs->vs_alloc);
+		/* size         5 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[5], vs->vs_space);
+		/* read_bytes   6 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[6], vs->vs_bytes[ZIO_TYPE_READ]);
+		/* iread_errors 7 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[7], vs->vs_read_errors);
+		/* read_ops     8 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[8], vs->vs_ops[ZIO_TYPE_READ]);
+		/* write_bytes  9 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[9], vs->vs_bytes[ZIO_TYPE_WRITE]);
+		/* write_errors 10 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[10], vs->vs_write_errors);
+		/* write_ops    11 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[11], vs->vs_ops[ZIO_TYPE_WRITE]);
+		/* checksum errors 12 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[12], vs->vs_checksum_errors);
+		/* fragmentation 13 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[13], vs->vs_fragmentation);
+		/* initialization errors 14 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[14], vs->vs_initialize_errors);
+	}
+
+        return (rc);
+}
+
+
+/*
+ * recursive stats printer
+ */
+static int get_recursive_stats(stat_printer_f func, nvlist_t *nvroot,
+    const char *pool_name, const char *parent_name, int descend)
+{
+        uint_t c, children;
+        nvlist_t **child;
+        char vdev_name[256];
+        int err;
+
+        /*err = func(nvroot, pool_name, parent_name);
+        if (err)
+                return (err);
+	*/
+        if (descend && nvlist_lookup_nvlist_array(nvroot, ZPOOL_CONFIG_CHILDREN,
+            &child, &children) == 0) {
+                (void) strlcpy(vdev_name, get_vdev_name(nvroot, parent_name),
+                    sizeof (vdev_name));
```

Comment:
Rather than generate the "vdevname" based on the internal ZFS representation, i.e. `root/draid-0/disk-17`.  I think it would be more user friendly to report names which align with the `zpool` command output.  In fact, I'd recommend we output three name columns, these can be generated with the `zpool_vdev_name()` library function.

1) `vdevname` - The short name reported by `zpool status` without any arguments.
    Command: `zpool status`
    Example: `U1`
    Flags: `None`
    
2) `vdevpath` - The full path to the underlying block device.
    Command:  `zpool status -PL`
    Example: `/dev/dm-156`
    Flags: `VDEV_NAME_PATH | VDEV_NAME_FOLLOW_LINKS`
   
3) `vdevguid` - The unique unsigned 64-bit guid assigned to the vdev.
    Command: `zpool status -g`
    Example: `4228566300096018771`
    Flags: `VDEV_NAME_GUID`
    
For the names of top-level vdevs this same function can be used with the `VDEV_NAME_TYPE_ID` flag.

Reply:
sounds good.